zio-schema/shared/src/main/scala/zio/schema/Diff.scala

```diff
@@ -413,427 +419,324 @@ object Differ {
   def enumN[Z](cases: Schema.Case[_, Z]*): Differ[Z] =
     (thisZ: Z, thatZ: Z) =>
       cases
-        .foldRight[Option[Diff]](None) {
+        .foldRight[Option[Diff[Z]]](None) {
           case (_, diff @ Some(_)) => diff
-          case (subtype, _) =>
+          case (subtype: Schema.Case[a, Z], _) =>
             subtype.deconstruct(thisZ) -> (subtype.deconstruct(thatZ)) match {
               case (Some(thisA), Some(thatA)) =>
-                Some(fromSchema(subtype.codec)(thisA, thatA))
+                val subtypeDiffer: Differ[Z] =
+                  fromSchema(subtype.codec).transform(
+                    (z: Z) => subtype.unsafeDeconstruct(z),
+                    (a: a) => a.asInstanceOf[Z]
+                  )
+                Some(subtypeDiffer(thisA.asInstanceOf[Z], thatA.asInstanceOf[Z]))
               case _ => None
             }
         }
-        .getOrElse(Diff.NotComparable)
+        .getOrElse(Diff.notComparable)
 
   def enumeration(structure: ListMap[String, Schema[_]]): Differ[(String, _)] =
     instancePartial[(String, _)] {
       case ((thisKey, thisValue), (thatKey, thatValue)) if thisKey == thatKey =>
         structure
           .get(thisKey)
-          .map(fromSchema(_).asInstanceOf[Differ[Any]].apply(thisValue, thatValue))
-          .getOrElse(Diff.NotComparable)
+          .map(fromSchema(_).asInstanceOf[Differ[Any]](thisValue, thatValue))
+          .getOrElse(Diff.notComparable)
+          .asInstanceOf[Diff[(String, Any)]]
     }
 
-  val string: Differ[String] = MyersDiff
+  val string: Differ[String] = LCSDiff.string
 
-  def instancePartial[A](f: PartialFunction[(A, A), Diff]): Differ[A] =
+  def instancePartial[A](f: PartialFunction[(A, A), Diff[A]]): Differ[A] =
     new Differ[A] {
-      override def apply(thisValue: A, thatValue: A): Diff =
-        f.applyOrElse((thisValue, thatValue), (_: (A, A)) => Diff.NotComparable)
+      override def apply(thisValue: A, thatValue: A): Diff[A] =
+        f.applyOrElse((thisValue, thatValue), (_: (A, A)) => Diff.notComparable)
     }
 
 }
 
-sealed trait Diff { self =>
+sealed trait Diff[A] { self =>
```

Comment:
I quite like the addition of this type parameter. And we don't lose anything, really, since the "untyped" version is just `Diff[_]`. ❤️ 👍